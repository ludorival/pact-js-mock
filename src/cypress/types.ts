export type HeadersConfig = {
  includes?: string[]
  excludes?: string[]
}
export type HeaderType = Record<string, string | string[]> | undefined
