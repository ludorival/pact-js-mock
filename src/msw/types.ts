/* eslint-disable @typescript-eslint/no-explicit-any */
import type { GraphQLError } from 'graphql'

export type GraphQLPayload = Record<string, any>
export type GraphQLResponse<T = any> = {
  data?: T
  errors?: GraphQLError[]
}
