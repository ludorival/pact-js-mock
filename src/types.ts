import type { GraphQLError } from 'graphql'
import type {
  DefaultBodyType,
  GraphQLRequest,
  GraphQLVariables,
  MockedRequest,
  Path,
  rest,
  RestRequest,
} from 'msw'

export type Body = DefaultBodyType
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type GraphQLPayload = Record<string, any>
export type GraphQLResponse<T> = {
  data?: T
  errors?: GraphQLError[]
}
export type Method = keyof typeof rest
export type AnyInteraction<
  RequestBodyType extends Body = Body,
  ResponseBodyType extends Body = Body,
  Variables extends GraphQLVariables = GraphQLVariables,
  GraphQLDataType extends GraphQLPayload = GraphQLPayload
> =
  | GraphQLInteraction<Variables, GraphQLDataType>
  | RestInteraction<RequestBodyType, ResponseBodyType>

export type CommonRecordInteraction = {
  description: string
  providerState?: string
  responseStatus?: number
  once?: boolean
  matchingRules?: MatchingRules
}

export type InterceptRequest = (
  request: MockedRequest | GraphQLRequest<GraphQLVariables>,
  response: Body
) => Promise<void>

export type RestInteraction<
  RequestBodyType extends Body = Body,
  ResponseBodyType extends Body = Body
> = CommonRecordInteraction & {
  api: 'rest'
  path: Path
  method: Method
  matchRequest?: (req: RestRequest<RequestBodyType>) => boolean
  response: ResponseBodyType
}
export type GraphQLInteraction<
  RequestBodyType extends GraphQLVariables = GraphQLVariables,
  ResponseBodyType extends GraphQLPayload = GraphQLPayload
> = CommonRecordInteraction & {
  api: 'graphql'
  type: 'query' | 'mutation'
  name: string
  matchRequest?: (req: GraphQLRequest<RequestBodyType>) => boolean
  data?: ResponseBodyType
  errors?: GraphQLError[]
}
export type HeadersConfig = {
  includes?: string[]
  excludes?: string[]
}
export type Pact = {
  consumer: string
  provider: string
  folder?: string
  headersConfig?: HeadersConfig
  basePath?: string
  pactSpecificationVersion?: '2.0.0'
}

export type HeaderType = Record<string, string | string[]> | undefined

export type Rule =
  | {
      match: 'type'
    }
  | { match: 'regexp'; regex: RegExp }

export type MatchingRules = { [key: string]: Rule }
type BaseXHR = {
  headers?: HeaderType
  body?: Body
}
export type Interaction = {
  description: string
  providerState?: string
  request: {
    method: string
    path: string
    query?: string
  } & BaseXHR
  response: {
    status: string | number | undefined
    matchingRules?: MatchingRules
  } & BaseXHR
}
export type MetaData = {
  pactSpecification: {
    version: string
  }
  client: {
    name: 'pact-msw-handlers'
    version: string
  }
}
export type PactFile = {
  consumer: { name: string }
  provider: { name: string }
  interactions: Interaction[]
  metadata: MetaData
}
