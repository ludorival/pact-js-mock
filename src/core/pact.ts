import { omit } from 'lodash'
import {
  HeaderType,
  HeadersConfig,
  InputPact,
  InteractionFor,
  MinimalInteraction,
  Options,
  PactFile,
  PactV2,
  PactV3,
  PactV4,
  Request,
  ToRecordInteraction,
  Version,
} from '../types'
import packageJson from '../../package.json'

function pactName<P extends PactFile>(pact: InputPact<P>) {
  return `${pact.consumer.name}-${pact.provider.name}`
}
export class Pact<T extends PactFile = PactV2.PactFile> {
  private interactions: InteractionFor<T>[] = []
  constructor(
    private pact: InputPact<T>,
    private options?: Options,
  ) {
    if (!pact.metadata?.pactSpecification?.version)
      throw new Error(`The version is missing in the Pact. Please provide the right version like
    { consumer: { name : 'my-consumer'}, provider : { name: 'my-provider'},
    metadata : {pactSpecification: { version : '2.0.0'}}}`)
  }

  public get name(): string {
    return pactName(this.pact)
  }
  public get version(): Version {
    return this.pact.metadata?.pactSpecification?.version || '2.0.0'
  }

  public get fileName(): string {
    return `${this.options?.outputDir || 'pacts'}/${this.name}.json`
  }

  record<TResponse = unknown, TRequest = unknown>(
    input: ToRecordInteraction<InteractionFor<T, TResponse, TRequest>>,
  ) {
    const request = input.request as Request
    const response = input.response as { status?: number }
    const description =
      input.description ||
      `${input.request?.method} ${input.request?.path} returns status ${
        response.status || 200
      }`

    const interaction = {
      description,
      ...input,
      request: { ...this.toRequest(request) },
    } as InteractionFor<T, TResponse, TRequest>
    const existingInteractionIndex = this.interactions.findIndex(
      (i) => i.description === description,
    )
    const existingInteraction = this.interactions[existingInteractionIndex]
    const sameInteraction =
      existingInteraction &&
      JSON.stringify(existingInteraction) == JSON.stringify(interaction)

    if (
      this.options?.deterministic &&
      existingInteraction &&
      !sameInteraction
    ) {
      throw new Error(
        `The interaction \`${description}\` already exists but with different content. It is recommended that the interaction stays deterministic.`,
      )
    }
    if (!existingInteraction) {
      this.interactions.push(interaction)
    } else {
      this.interactions[existingInteractionIndex] = interaction
    }
  }

  private toRequest(req: Request): Request {
    const { headersConfig, basePath } = this.options || {}
    const path = req.path?.replace(basePath || '', '')
    if (req.headers) {
      req.headers = omitHeaders(req.headers, headersConfig)
    }
    return {
      ...req,
      path,
    } as Request
  }

  generatePactFile(): T {
    const pactFile = {
      ...(this.pact as T),
      interactions: this.interactions,
      metadata: {
        ...this.pact.metadata,
        client: { name: packageJson.name, version: packageJson.version },
      },
    }
    return pactFile
  }

  async reset(pactFile: T | null = null) {
    this.interactions = (pactFile?.interactions || []) as InteractionFor<T>[]
  }
}

const omitHeaders = (
  headers: HeaderType,
  headersConfig: HeadersConfig = {},
) => {
  const blocklist = headersConfig.excludes || []
  if (headersConfig.includes) {
    const remove = Object.keys(omit(headers, ...headersConfig.includes))
    blocklist.push(...remove)
  }
  return omit(headers, [...blocklist])
}

export function buildResponse<T>(content: T, version: Version) {
  switch (version) {
    case '2.0.0':
      return {
        response: { status: 200, body: content },
      } as MinimalInteraction<PactV2.Interaction>
    case '3.0.0':
      return {
        response: { status: 200, body: content },
      } as MinimalInteraction<PactV3.Interaction>

    case '4.0.0':
      return {
        response: {
          status: 200,
          body: { content, contentType: 'application/json' },
        },
      } as MinimalInteraction<PactV4.Interaction>
  }
}
