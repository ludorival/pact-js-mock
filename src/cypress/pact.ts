import type {
  CyHttpMessages,
  RouteHandlerController,
  StaticResponse,
} from 'cypress/types/net-stubbing'
import { Pact as BasePact, buildResponse } from '../core'
import {
  InputPact,
  InteractionFor,
  MinimalInteraction,
  PactFile,
  PactV2,
  PactV3,
  PactV4,
  ToRecordInteraction,
} from '../types'
import { HeadersConfig } from './types'

type Options = {
  headersConfig?: HeadersConfig
  basePath?: string
}

type Request = PactV2.Request | PactV3.Request | PactV4.Request
export class Pact<P extends PactFile> extends BasePact<P> {
  constructor(pact: InputPact<P>, options?: Options) {
    super(pact, options)
  }

  toHandler<T extends object>(
    input: MinimalInteraction<InteractionFor<P, T>> | T
  ): RouteHandlerController {
    const version = this.version
    return (req) => {
      const interaction =
        'response' in input ? input : buildResponse(input, version)
      req.reply(
        this.recordResponse(
          interaction as MinimalInteraction<InteractionFor<P, T>>,
          req
        )
      )
    }
  }
  recordResponse(
    interaction: MinimalInteraction<InteractionFor<P>>,
    req: CyHttpMessages.IncomingHttpRequest
  ): StaticResponse {
    const request = toRequest(req)
    this.record({ ...interaction, request } as ToRecordInteraction<
      InteractionFor<P>
    >)
    const responseV4 = (interaction.response as { body?: { content: unknown } })
      .body?.content
      ? (interaction.response as PactV4.ResponseClass)
      : null
    const response = responseV4
      ? null
      : (interaction.response as PactV2.Response | PactV3.Response)

    return {
      statusCode: responseV4?.status || response?.status,
      body: responseV4?.body || response?.body,
      headers: responseV4?.headers || response?.headers,
    }
  }
}

function toRequest(req: CyHttpMessages.IncomingHttpRequest): Request {
  const parts = req.url.split('?')
  const path = parts[0]
  const query = parts.slice(1).join('?')
  const body: Body = req.body

  return {
    method: req.method,
    path,
    headers: req.headers,
    body,
    query,
  } as Request
}
