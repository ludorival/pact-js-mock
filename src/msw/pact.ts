import { omit } from 'lodash'
import {
  DefaultBodyType,
  HttpResponse,
  HttpResponseInit,
  StrictRequest,
  StrictResponse,
} from 'msw'
import { Pact as BasePact, buildResponse } from '../core'
import {
  InputPact,
  InteractionFor,
  MinimalInteraction,
  Options,
  PactFile,
  PactV2,
  PactV3,
  PactV4,
  Request,
  ToRecordInteraction,
} from '../types'
type Info<
  R extends DefaultBodyType = DefaultBodyType,
  Extra extends Record<string, unknown> = Record<string, unknown>,
> = {
  request: StrictRequest<R>
} & Extra
export class Pact<P extends PactFile> extends BasePact<P> {
  constructor(pact: InputPact<P>, options?: Options) {
    super(pact, options)
  }

  toResolver<T extends object, R extends DefaultBodyType = DefaultBodyType>(
    input: MinimalInteraction<InteractionFor<P, T>> | T,
  ): (info: Info<R>) => Promise<StrictResponse<T>> {
    const version = this.version
    return async (info) => {
      const interaction =
        'response' in input ? input : buildResponse(input, version)
      const response = await this.toResponse(
        interaction as MinimalInteraction<InteractionFor<P, T>>,
        info,
      )
      return response
    }
  }

  async toResponse<
    TResponse extends DefaultBodyType,
    TRequest extends DefaultBodyType = DefaultBodyType,
  >(
    interaction: MinimalInteraction<InteractionFor<P, TResponse>>,
    info: Info<TRequest>,
    initOptions?: HttpResponseInit,
  ): Promise<StrictResponse<TResponse>> {
    const toRecord = {
      ...interaction,
      request: await toRequest(info),
    } as ToRecordInteraction<InteractionFor<P>>
    this.record(toRecord)

    const responseV4 = (interaction.response as { body?: { content: unknown } })
      .body?.content
      ? (interaction.response as PactV4.ResponseClass)
      : null
    const response = responseV4
      ? null
      : (interaction.response as PactV2.Response | PactV3.Response)
    const content = responseV4?.body?.content || response?.body

    const responseStatus = responseV4?.status || response?.status || 200

    return HttpResponse.json(content, {
      headers: response?.headers || responseV4?.headers,
      status: responseStatus,
      ...initOptions,
    })
  }
}

async function toRequest<R extends DefaultBodyType = DefaultBodyType>(
  info: Info<R>,
): Promise<Request> {
  const url = new URL(info.request.url)
  const path = url.pathname
  const query = url.searchParams.toString() || undefined
  const body = info.query
    ? omit(info, 'request', 'requestId')
    : await info.request.json().catch(() => info.request.body)
  const headers: Record<string, unknown> = {}
  info.request.headers.forEach((value, key) => (headers[key] = value))
  return {
    method: info.request.method,
    path,
    headers,
    body,
    query,
  } as Request
}
