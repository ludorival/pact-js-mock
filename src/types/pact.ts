import * as PactV2 from './pactv2'
import * as PactV3 from './pactv3'
import * as PactV4 from './pactv4'

export type HeaderType = Record<string, string | string[]> | undefined

export type PactFile = PactV2.PactFile | PactV3.PactFile | PactV4.PactFile

export type Interaction<TResponse = unknown> =
  | PactV2.Interaction<TResponse>
  | PactV3.Interaction<TResponse>
  | PactV4.Interaction<TResponse>

export type MinimalInteraction<I extends Interaction = PactV2.Interaction> =
  Partial<I> & Pick<I, 'response'>

export type ToRecordInteraction<I extends Interaction> = Partial<Partial<I>> &
  Pick<I, 'response' | 'request'>

export type Version =
  | PactV2.PactSpecification['version']
  | PactV3.PactSpecification['version']
  | PactV4.PactSpecification['version']

export type InteractionFor<
  P extends PactFile,
  TResponse = unknown,
  TRequest = unknown,
> = P extends PactV2.PactFile
  ? PactV2.Interaction<TResponse, TRequest>
  : P extends PactV3.PactFile
    ? PactV3.Interaction<TResponse, TRequest>
    : P extends PactV4.PactFile
      ? PactV4.Interaction<TResponse, TRequest>
      : never

export type InputPact<P extends PactFile> = Partial<P> &
  Pick<P, 'consumer' | 'provider' | 'metadata'>

export { PactV2, PactV3, PactV4 }

export type Request = PactV2.Request | PactV3.Request | PactV4.Request

export type Options = {
  headersConfig?: HeadersConfig
  basePath?: string
  deterministic?: boolean
}

export type HeadersConfig = {
  includes?: string[]
  excludes?: string[]
}
