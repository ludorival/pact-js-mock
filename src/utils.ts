import { omit } from 'lodash'
import { MockedRequest } from 'msw'
import path from 'path'
import {
  Body,
  HeadersConfig,
  HeaderType,
  Interaction,
  Pact,
  PactFile,
} from './types'
const omitHeaders = (
  headers: HeaderType,
  headersConfig: HeadersConfig = {}
) => {
  const blocklist = headersConfig.excludes || []
  if (headersConfig.includes) {
    const remove = Object.keys(omit(headers, ...headersConfig.includes))
    blocklist.push(...remove)
  }
  return omit(headers, [...blocklist])
}

export const toRequest = async (
  req: MockedRequest,
  headersConfig?: HeadersConfig
) => {
  const path = req.url.pathname
  const query = req.url.searchParams.toString() || undefined
  const body: Body = await req
    .json()
    .catch(() => req.text())
    .catch(() => undefined)

  return {
    method: req.method,
    path,
    headers: omitHeaders(req.headers.all(), headersConfig),
    body,
    query,
  } as Interaction['request']
}
type FileSystem = {
  existsSync: (filename: string) => boolean
  mkdirSync: (dirname: string) => void
  writeFileSync: (filename: string, data: string) => void
  readFileSync: (filename: string) => Buffer
  rmSync: (filename: string, options: { recursive: boolean }) => void
}
let fs: FileSystem
const checkIfNodeEnv = (): FileSystem | undefined => {
  if (!fs) {
    try {
      fs = require('fs')
    } catch (e) {}
  }
  if (!fs?.existsSync) {
    console.warn('You need a node environment to save files.', {
      mode: 'warning',
      group: true,
    })
    return undefined
  } else {
    return fs
  }
}
const ensureDirExists = (filePath: string) => {
  const dirname = path.dirname(filePath)
  if (fs.existsSync?.(dirname)) {
    return true
  }
  fs.mkdirSync?.(dirname)
}

export const deleteFile = (filePath: string) => {
  const fs = checkIfNodeEnv()
  if (fs?.existsSync?.(filePath)) {
    fs.rmSync(filePath, { recursive: true })
  }
}
export const readPact = (filePath: string): PactFile | undefined => {
  const fs = checkIfNodeEnv()
  if (fs) {
    if (!fs.existsSync(filePath)) return undefined
    try {
      return JSON.parse(fs.readFileSync(filePath).toString())
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } catch (e: any) {
      console.warn(`The pactFile ${filePath} might not exist: ${e.message}`)
      return undefined
    }
  }
}

export const writePact = (filePath: string, data: PactFile) => {
  const fs = checkIfNodeEnv()
  if (fs) {
    ensureDirExists(filePath)
    fs.writeFileSync?.(filePath, JSON.stringify(data, null, 2))
  }
}

export const ensureSamePact = (pact: Pact, pactFile?: PactFile) => {
  if (!pactFile) return
  if (
    pactFile?.consumer.name !== pact.consumer ||
    pactFile?.provider.name !== pact.provider
  )
    throw new Error(
      `The pactFile to load does not match with the expected couple (${pact.consumer}-${pact.provider} )`
    )
}

export const isNotTheSame = ({
  fileName,
  existingInteraction,
  interaction,
}: {
  fileName: string
  existingInteraction?: Interaction
  interaction: Interaction
}): boolean => {
  if (!existingInteraction) return true
  const isSame = existingInteraction
    ? JSON.stringify(existingInteraction) === JSON.stringify(interaction)
    : true

  if (!isSame) {
    throw new Error(
      `The pact ${fileName} contains an interaction with same description ${
        interaction.description
      }.
Previous interaction was 
${JSON.stringify(existingInteraction, null, 2)}
Current interaction is
${JSON.stringify(interaction, null, 2)}
----------------------------------
Make sure your mocks are deterministic. 
You can use headersConfig options from your provider
to exclude or include headers`
    )
  }
  return !isSame
}
