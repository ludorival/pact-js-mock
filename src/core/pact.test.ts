import { omitVersion } from '../test/utils'
import { PactV2, PactV3, PactV4 } from '../types'
import { deletePact, reloadPact, writePact } from '../utils'
import { Pact } from './pact'

const anInteraction = {
  description: 'This is a description',
  request: { method: 'POST', path: 'v1/todo', body: { name: 'Todo' } },
  response: {
    status: 200,
    body: 'This is an interaction with a description',
  },
} as PactV2.Interaction
describe('PactV2', () => {
  const pact = new Pact<PactV2.PactFile>(
    {
      consumer: { name: 'consumer' },
      provider: { name: 'provider' },
      metadata: { pactSpecification: { version: '2.0.0' } },
    },
    {
      headersConfig: {
        includes: ['content-type'],
        excludes: ['Authorization'],
      },
      basePath: 'base',
    },
  )

  afterEach(() => {
    pact.reset()
  })

  it('should record a Pact V2 Interactions ', () => {
    // when
    pact.record({
      request: { method: 'GET', path: 'v1/todo' },
      response: {
        status: 200,
        body: 'This is an interaction without description',
      },
    })
    pact.record(anInteraction)
    pact.record({
      ...anInteraction,
      description: 'This is a same description',
    })
    pact.record({
      ...anInteraction,
      description: 'This is a same description',
      request: { method: 'GET', path: 'v1/todo' },
    })
    pact.record({
      ...anInteraction,
      description: 'Description with special character @/"". $',
    })
    pact.record({
      ...anInteraction,
      description: 'This is an interaction with matching rules',
      request: {
        ...anInteraction.request,
        matchingRules: { '$.body': { match: 'type' } },
      },
    })
    pact.record({
      ...anInteraction,
      description: 'This is an interaction with headers and base path',
      request: {
        ...anInteraction.request,
        path: 'base/v1/todo',
        headers: {
          'content-type': 'application/json',
          Authorization: 'Bearer xxx',
        },
      },
    })

    const pactFile = pact.generatePactFile()
    expect(omitVersion(pactFile)).toMatchSnapshot()
  })
})

describe('PactV3', () => {
  const pact = new Pact<PactV3.PactFile>({
    consumer: { name: 'consumer' },
    provider: { name: 'provider' },
    metadata: { pactSpecification: { version: '3.0.0' } },
  })

  afterEach(() => {
    pact.reset()
  })

  it('should record a Pact V3 Interactions ', () => {
    // given

    const anInteraction = {
      description: 'This is a description',
      request: { method: 'POST', path: 'v1/todo', body: { name: 'Todo' } },
      response: {
        status: 200,
        body: 'This is an interaction with a description',
      },
    } as PactV3.Interaction

    // when
    pact.record({
      request: { method: 'GET', path: 'v1/todo' },
      response: {
        status: 200,
        body: 'This is an interaction without description',
      },
    })
    pact.record(anInteraction)
    pact.record({
      ...anInteraction,
      description: 'This is a same description',
    })
    pact.record({
      ...anInteraction,
      description: 'This is a same description',
      request: { method: 'GET', path: 'v1/todo' },
    })
    pact.record({
      ...anInteraction,
      description: 'Description with special character @/"". $',
    })
    pact.record({
      ...anInteraction,
      description: 'This is an interaction with matching rules',
      request: {
        ...anInteraction.request,
        matchingRules: { body: { match: 'type' } },
      },
    })

    const pactFile = pact.generatePactFile()
    expect(omitVersion(pactFile)).toMatchSnapshot()
  })
})

describe('PactV4', () => {
  const pact = new Pact<PactV4.PactFile>({
    consumer: { name: 'consumer' },
    provider: { name: 'provider' },
    metadata: { pactSpecification: { version: '4.0.0' } },
  })

  afterEach(() => {
    pact.reset()
  })

  it('should record a Pact V4 Interactions ', () => {
    // given

    const anInteraction = {
      type: 'Synchronous/HTTP',
      description: 'This is a description',
      request: {
        method: 'POST',
        path: 'v1/todo',
        body: { content: { name: 'Todo' }, contentType: 'application/json' },
      },
      response: {
        body: {
          content: 'This is an interaction with a description',
          contentType: 'application/json',
        },
      },
    } as PactV4.Interaction

    // when
    pact.record({
      request: { method: 'GET', path: 'v1/todo' },
      response: {
        status: 200,
        body: {
          content: 'This is an interaction without description',
          contentType: 'application/json',
        },
      },
    })
    pact.record(anInteraction)
    pact.record({
      ...anInteraction,
      description: 'This is a same description',
    })
    pact.record({
      ...anInteraction,
      description: 'This is a same description',
      request: { method: 'GET', path: 'v1/todo' },
    })
    pact.record({
      ...anInteraction,
      description: 'Description with special character @/"". $',
    })
    pact.record({
      ...anInteraction,
      description: 'This is an interaction with matching rules',
      request: {
        ...anInteraction.request,
        matchingRules: { body: { match: 'type' } },
      },
    })
    pact.record({
      ...anInteraction,
      description: 'This is an interaction with undefined body',
      response: {
        status: 403,
        body: undefined,
      },
    })

    const pactFile = pact.generatePactFile()
    expect(omitVersion(pactFile)).toMatchSnapshot()
  })
})

describe('Deterministic Pact', () => {
  // given
  const pact = new Pact(
    {
      consumer: { name: 'consumer' },
      provider: { name: 'provider' },
      metadata: { pactSpecification: { version: '2.0.0' } },
    },
    {
      deterministic: true,
    },
  )

  beforeEach(() => {
    pact.reset()
  })
  it('should not be able to record an interaction with same description but different request when deterministic option is set', () => {
    // when
    pact.record(anInteraction)
    // then
    expect(() =>
      pact.record({
        ...anInteraction,
        request: { method: 'POST', path: 'v1/todo', body: { name: 'Todo 2' } },
      }),
    ).toThrow(
      'The interaction `This is a description` already exists but with different content. It is recommended that the interaction stays deterministic.',
    )
  })

  it('should be able to record same interactions multiple times', () => {
    // when
    pact.record(anInteraction)
    pact.record(anInteraction)
    pact.record(anInteraction)
    // then
    expect(pact.generatePactFile().interactions).toHaveLength(1)
  })
})

describe('Reload pacts', () => {
  const pact = new Pact(
    {
      consumer: { name: 'consumer' },
      provider: { name: 'provider' },
      metadata: { pactSpecification: { version: '2.0.0' } },
    },
    {
      outputDir: 'pacts/test-reload',
    },
  )
  beforeAll(() => {
    deletePact(pact)
  })
  beforeEach(() => {
    reloadPact(pact)
  })

  afterEach(() => {
    writePact(pact)
  })

  it('should record an interaction', () => {
    pact.record(anInteraction)
    expect(pact.generatePactFile().interactions).toHaveLength(1)
  })

  it('should keep the previous interaction', () => {
    pact.record({
      ...anInteraction,
      description: 'This is another description',
    })
    expect(pact.generatePactFile().interactions).toHaveLength(2)
  })
})
