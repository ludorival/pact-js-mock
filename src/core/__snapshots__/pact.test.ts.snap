// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PactV2 should record a Pact V2 Interactions  1`] = `
{
  "consumer": {
    "name": "consumer",
  },
  "interactions": [
    {
      "description": "GET v1/todo returns status 200",
      "request": {
        "method": "GET",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction without description",
        "status": 200,
      },
    },
    {
      "description": "This is a description",
      "request": {
        "body": {
          "name": "Todo",
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction with a description",
        "status": 200,
      },
    },
    {
      "description": "This is a same description",
      "request": {
        "method": "GET",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction with a description",
        "status": 200,
      },
    },
    {
      "description": "Description with special character @/"". $",
      "request": {
        "body": {
          "name": "Todo",
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction with a description",
        "status": 200,
      },
    },
    {
      "description": "This is an interaction with matching rules",
      "request": {
        "body": {
          "name": "Todo",
        },
        "matchingRules": {
          "$.body": {
            "match": "type",
          },
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction with a description",
        "status": 200,
      },
    },
    {
      "description": "This is an interaction with headers and base path",
      "request": {
        "body": {
          "name": "Todo",
        },
        "headers": {
          "content-type": "application/json",
        },
        "method": "POST",
        "path": "/v1/todo",
      },
      "response": {
        "body": "This is an interaction with a description",
        "status": 200,
      },
    },
  ],
  "metadata": {
    "client": {
      "name": "pact-js-mock",
    },
    "pactSpecification": {
      "version": "2.0.0",
    },
  },
  "provider": {
    "name": "provider",
  },
}
`;

exports[`PactV3 should record a Pact V3 Interactions  1`] = `
{
  "consumer": {
    "name": "consumer",
  },
  "interactions": [
    {
      "description": "GET v1/todo returns status 200",
      "request": {
        "method": "GET",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction without description",
        "status": 200,
      },
    },
    {
      "description": "This is a description",
      "request": {
        "body": {
          "name": "Todo",
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction with a description",
        "status": 200,
      },
    },
    {
      "description": "This is a same description",
      "request": {
        "method": "GET",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction with a description",
        "status": 200,
      },
    },
    {
      "description": "Description with special character @/"". $",
      "request": {
        "body": {
          "name": "Todo",
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction with a description",
        "status": 200,
      },
    },
    {
      "description": "This is an interaction with matching rules",
      "request": {
        "body": {
          "name": "Todo",
        },
        "matchingRules": {
          "body": {
            "match": "type",
          },
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction with a description",
        "status": 200,
      },
    },
  ],
  "metadata": {
    "client": {
      "name": "pact-js-mock",
    },
    "pactSpecification": {
      "version": "3.0.0",
    },
  },
  "provider": {
    "name": "provider",
  },
}
`;

exports[`PactV4 should record a Pact V4 Interactions  1`] = `
{
  "consumer": {
    "name": "consumer",
  },
  "interactions": [
    {
      "description": "GET v1/todo returns status 200",
      "request": {
        "method": "GET",
        "path": "v1/todo",
      },
      "response": {
        "body": {
          "content": "This is an interaction without description",
          "contentType": "application/json",
        },
        "status": 200,
      },
    },
    {
      "description": "This is a description",
      "request": {
        "body": {
          "content": {
            "name": "Todo",
          },
          "contentType": "application/json",
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": {
          "content": "This is an interaction with a description",
          "contentType": "application/json",
        },
      },
      "type": "Synchronous/HTTP",
    },
    {
      "description": "This is a same description",
      "request": {
        "method": "GET",
        "path": "v1/todo",
      },
      "response": {
        "body": {
          "content": "This is an interaction with a description",
          "contentType": "application/json",
        },
      },
      "type": "Synchronous/HTTP",
    },
    {
      "description": "Description with special character @/"". $",
      "request": {
        "body": {
          "content": {
            "name": "Todo",
          },
          "contentType": "application/json",
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": {
          "content": "This is an interaction with a description",
          "contentType": "application/json",
        },
      },
      "type": "Synchronous/HTTP",
    },
    {
      "description": "This is an interaction with matching rules",
      "request": {
        "body": {
          "content": {
            "name": "Todo",
          },
          "contentType": "application/json",
        },
        "matchingRules": {
          "body": {
            "match": "type",
          },
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": {
          "content": "This is an interaction with a description",
          "contentType": "application/json",
        },
      },
      "type": "Synchronous/HTTP",
    },
    {
      "description": "This is an interaction with undefined body",
      "request": {
        "body": {
          "content": {
            "name": "Todo",
          },
          "contentType": "application/json",
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": undefined,
        "status": 403,
      },
      "type": "Synchronous/HTTP",
    },
  ],
  "metadata": {
    "client": {
      "name": "pact-js-mock",
    },
    "pactSpecification": {
      "version": "4.0.0",
    },
  },
  "provider": {
    "name": "provider",
  },
}
`;
